index.js

```diff
@@ -125,6 +152,168 @@
+    return t.type == 'UNARY' || t.type == 'BINARY';
+  }
+  
+  var sigArgument = function(t, offset, isChild) {
+    var item = sigItem(t.type, offset,
+                       t.name ? stripNamespace(t.name) : t.toString(), ' ');
+    if (t.type == 'VARIABLE') {
+      item.typeVars[t.name] = sigTypeVar(t.name, offset);
+    } else if (isContainerType(t)) {
+      if (isChild) {
+        item.text = '(' + item.text;
+        item.end += 1;
+      }
+      var childTypes = [t.$1];
+      if (t.type == 'BINARY') {
+        childTypes.push(t.$2);
+      }
+      var idx, child;
+      for (idx = 0; idx < childTypes.length; idx += 1) {
+        item.end += item.glue.length;
+        child = sigArgument(childTypes[idx], item.end, true);
+        if (isContainerType(child)) {
+          item.children.push(child);
+        } else {
+          item.typeVars = merge(item.typeVars)(child.typeVars);
+        }
+        item.text += item.glue + child.text;
+        item.end = offset + item.text.length;
+      }
+      if (isChild) {
+        item.text += ')';
+        item.end += 1;
+      }
+    }
+    return item;
+  };
+  
+  var sigReturnType = function(expRetType, offset) {
+    var group = sigGroup(offset, '', '');
+    var item = sigArgument(expRetType, offset, false);
+    group.end = item.end;
+    group.items.push(item);
+    return group;
+  };
+  
+  var strRepeat = function(s, times) {
+    return Array(times + 1).join(s);
+  };
+  
+  var sigArgumentFindTypeVar = function(name) {
+    return function recur(item) {
+        var positions = name in item.typeVars ? [item.typeVars[name].offset] :
+                                                [];
+        return positions.concat(chain(item.children, recur));
+    };
+  };
+  
+  var sigShowTypeVarPositions = function(sig, name) {
+    var idx;
+    var positions = chain(sig.arguments.items, sigArgumentFindTypeVar(name));
+    
+    var upcarrots = '';
+    for (idx = 0; idx < positions.length; idx += 1) {
+      var pos = positions[idx];
+      upcarrots += strRepeat(' ', pos - upcarrots.length) + '^';
+    }
+    var varNumbers = '';
+    for (idx = 0; idx < positions.length; idx += 1) {
+      var pos = positions[idx];
+      varNumbers += strRepeat(' ', pos - varNumbers.length) + (idx + 1);
+    }
+    return upcarrots + '\n' + varNumbers;
+  };
+  
+  //  signature :: (String -> [{String: [TypeClass]}] -> [Type] -> Type) -> Object
+  var signature = function(name, constraints, expTypes, expRetType) {
+    var sig = {};
+    sig.name = sigName(name, 0);
+    sig.constraints = sigConstraints(constraints, sig.name.end);
+    sig.arguments = sigArguments(expTypes, sig.constraints.end);
+    sig.returnType = sigReturnType(expRetType, sig.arguments.end);
+    sig.toString = function() {
+      return (
+        sigGroupStr(sig.name) +
+        sigGroupStr(sig.constraints) +
+        sigGroupStr(sig.arguments) +
+        sigGroupStr(sig.returnType)
+      );
+    };
+    return sig;
+  };
```

Comment:
I've been working on an approach which seems simpler to me.

```javascript
//  literal :: String -> Token
var literal = function(text) {
  return {type: 'LITERAL', text: text};
};

//  variable :: String -> Token
var variable = function(name) {
  return {type: 'VARIABLE', name: name};
};

//  group :: String,String,String -> [Token] -> Token
var group = function(open, separator, close) {
  return function(tokens) {
    return {type: 'GROUP',
            open: open,
            separator: separator,
            close: close,
            tokens: tokens};
  };
};

var EMPTY = '';
var SPACE = ' ';

var constraints = group('(', ', ', ')');
var constraint  = group(EMPTY, SPACE, EMPTY);
var types       = group(EMPTY, ' -> ', EMPTY);
var type        = group('(', SPACE, ')');

//  serialize :: Token -> String
var serialize = function recur(token) {
  switch (token.type) {
    case 'LITERAL':
      return token.text;
    case 'VARIABLE':
      return token.name;
    default:
      return token.open +
             token.tokens.map(recur).join(token.separator) +
             token.close;
  }
};

//  token :: Token
var token =
group(EMPTY, ' :: ', EMPTY)([
  literal('Either.concat'),
  group(EMPTY, ' => ', EMPTY)([
    constraints([constraint([literal('Semigroup'), variable('a')]),
                 constraint([literal('Semigroup'), variable('b')])]),
    types([type([literal('Either'), variable('a'), variable('b')]),
           type([literal('Either'), variable('a'), variable('b')]),
           type([literal('Either'), variable('a'), variable('b')])]),
  ]),
]);

token;
// => { type: 'GROUP',
//      open: '',
//      separator: ' :: ',
//      close: '',
//      tokens:
//       [ { type: 'LITERAL', text: 'Either.concat' },
//         { type: 'GROUP',
//           open: '',
//           separator: ' => ',
//           close: '',
//           tokens:
//            [ { type: 'GROUP',
//                open: '(',
//                separator: ', ',
//                close: ')',
//                tokens:
//                 [ { type: 'GROUP',
//                     open: '',
//                     separator: ' ',
//                     close: '',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Semigroup' },
//                        { type: 'VARIABLE', name: 'a' } ] },
//                   { type: 'GROUP',
//                     open: '',
//                     separator: ' ',
//                     close: '',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Semigroup' },
//                        { type: 'VARIABLE', name: 'b' } ] } ] },
//              { type: 'GROUP',
//                open: '',
//                separator: ' -> ',
//                close: '',
//                tokens:
//                 [ { type: 'GROUP',
//                     open: '(',
//                     separator: ' ',
//                     close: ')',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Either' },
//                        { type: 'VARIABLE', name: 'a' },
//                        { type: 'VARIABLE', name: 'b' } ] },
//                   { type: 'GROUP',
//                     open: '(',
//                     separator: ' ',
//                     close: ')',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Either' },
//                        { type: 'VARIABLE', name: 'a' },
//                        { type: 'VARIABLE', name: 'b' } ] },
//                   { type: 'GROUP',
//                     open: '(',
//                     separator: ' ',
//                     close: ')',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Either' },
//                        { type: 'VARIABLE', name: 'a' },
//                        { type: 'VARIABLE', name: 'b' } ] } ] } ] } ] }

serialize(token);
// => 'Either.concat :: (Semigroup a, Semigroup b) => (Either a b) -> (Either a b) -> (Either a b)'
```

Outstanding tasks:

  - Avoid including `'() => '` when there are no constraints. I believe a general solution exists.
  - Generate the `^` markers from the same data structure.
  - Write a function to generate a value of type `Token` from `(name, constraints, expTypes, expRetType)`.